### PR TITLE
Adding lazy relation -> data.frame conversion for R client

### PR DESCRIPTION
**Pull request by hannes**
We add functionality to convert a R DuckDB relation object to a data.frame without actually computing the query result using lazy ALTREP vectors. This works great except for list/struct/blob columns. There is hope R will add list ALTREP support soon-ish, which should remove the limitation.

### CODE REVIEW

File: tools/rpkg/src/reltoaltrep.cpp [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+	shared_ptr<Relation> rel;
+	unique_ptr<QueryResult> res;
+};
+
+struct AltrepRownamesWrapper {
+
+	AltrepRownamesWrapper(shared_ptr<AltrepRelationWrapper> rel_p) : rel(rel_p) {
+		rowlen_data[0] = NA_INTEGER;
+		rowlen_data[1] = -42;
```
Comment (krlmlr) [inline tools/rpkg/src/reltoaltrep.cpp:83]: Do we need this?

```suggestion
```

File: tools/rpkg/src/reltoaltrep.cpp [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+static SEXP install_new_attrib(SEXP vec, SEXP name, SEXP val) {
+	SEXP attrib_vec = ATTRIB(vec);
+	SEXP attrib_cell = Rf_cons(val, R_NilValue);
+	SET_TAG(attrib_cell, name);
+	SETCDR(attrib_vec, attrib_cell);
+	return val;
+}
+
+R_xlen_t RelToAltrep::RownamesLength(SEXP x) {
+	return 2;
```
Comment (krlmlr) [inline tools/rpkg/src/reltoaltrep.cpp:152]: Magic number?

File: tools/rpkg/src/reltoaltrep.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	case LogicalTypeId::TIMESTAMP_NS:
+	case LogicalTypeId::DATE:
+	case LogicalTypeId::TIME:
+	case LogicalTypeId::INTERVAL:
+		return RelToAltrep::real_class;
+	case LogicalTypeId::VARCHAR:
+	case LogicalTypeId::UUID:
+		return RelToAltrep::string_class;
+	default:
+		cpp11::stop("rel_to_altrepUnknown column type for altrep: %s", type.ToString().c_str());
```
Comment (krlmlr) [inline tools/rpkg/src/reltoaltrep.cpp:207]:
```suggestion
		cpp11::stop("rel_to_altrep: Unknown column type for altrep: %s", type.ToString().c_str());
```

File: tools/rpkg/src/reltoaltrep.cpp [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	vector<string> names;
+	for (auto &col : rel->Columns()) {
+		names.push_back(col.Name());
+	}
+
+	SET_NAMES(data_frame, StringsToSexp(names));
+	for (size_t col_idx = 0; col_idx < ncols; col_idx++) {
+		cpp11::external_pointer<AltrepVectorWrapper> ptr(new AltrepVectorWrapper(relation_wrapper, col_idx));
+		auto column_type = rel->Columns()[col_idx].Type();
+		auto vector_sexp = r_protector.Protect(R_new_altrep(LogicalTypeToAltrepType(column_type), ptr, R_NilValue));
```
Comment (krlmlr) [inline tools/rpkg/src/reltoaltrep.cpp:233]: AFAICT protection not needed here?

```suggestion
		auto vector_sexp = R_new_altrep(LogicalTypeToAltrepType(column_type), ptr, R_NilValue);
```

**Review by krlmlr — APPROVED** [review]
🎉

File: tools/rpkg/src/reltoaltrep.cpp (hunk 4 above)
Comment (hannes) [reply tools/rpkg/src/reltoaltrep.cpp:235]: we now need it

**Comment by hannes** [comment]
@krlmlr do we need to remove the examples for the un-exported methods? R CMD check does not seem to like them.

**Comment by krlmlr** [comment]
I think so, yes.

**Comment by hannes** [comment]
I'm also wondering why those things have to be private in the first place. Maybe its better to mark them experimental.

**Comment by krlmlr** [comment]
We agreed that keeping the functions private will make it easier for us later to use the new relational package. What is the advantage of keeping them exported?

**Review by krlmlr — APPROVED** [review]
I don't have a strong opinion regarding reexport, I see there is now a message that explicitly alerts users. Would you like a beautiful orange badge, like in https://dplyr.tidyverse.org/reference/rows.html?

**Comment by hannes** [comment]
> Would you like a beautiful orange badge, like in https://dplyr.tidyverse.org/reference/rows.html?

Since this pulls in the lifecycle package I am adding `@noRd` as discussed

**Pull request merged**